File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-20.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        def isSametree(main, sub):
            if not main and not sub:
                return True

            if main and sub and main.val == sub.val:
                return isSametree(main.left, sub.left) and isSametree(main.right, sub.right)

            else:
                return False

        if not root:
            return False

        if not subRoot:
            return True

        if isSametree(root, subRoot):
            return True

        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-20.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (!root) {
            return {};
        }

        std::queue<TreeNode*> q;
        q.push(root);
        vector<vector<int>> res;

        while (q.size() > 0) {
            vector<int> level;
            int qlen = q.size();

            for (int i = 0; i < qlen; i++) {
                TreeNode* cur = q.front();
                q.pop();
                level.push_back(cur->val);

                if (cur->left != nullptr) {
                    q.push(cur->left);
                }
                if (cur->right != nullptr) {
                    q.push(cur->right);
                }
            } 

            if (!level.empty()) {
                res.push_back(level);
            }
            
        }

        return res;

    }
};
